File: python_code/memory_manager.py
import memory_primitives as mp
import sched
import time
import threading as th

class MemoryManager:
    def __init__(
        self,
        memory_range : tuple[int, int],
    ):
        self.memory_range = memory_range

        self.memory = {
            i: mp.MemoryItem(
                data=None,
                status="E",
            )
            for i in range(self.memory_range[0], self.memory_range[1])
        }

        self.locks = {
            i: mp.LockItem() for i in range(self.memory_range[0], self.memory_range[1])
        }

        self.copy_holders : dict[int, list[tuple[str, int]]] = {
            i: [] for i in range(self.memory_range[0], self.memory_range[1])
        }

    def read_memory(self, address: int) -> None | mp.MemoryItem:
        if address not in self.memory:
            return None
        return self.memory[address]
    
    def write_memory(self, address: int, data) -> None | mp.MemoryItem:
        if address not in self.memory:
            return None
        self.memory[address].data = data
        self.memory[address].wtag += 1
        return self.memory[address]
    
    def acquire_lock(self, address: int, lease_seconds=None) -> tuple[bool, int, int]:
        """
        Return:
        - ret_val: True if the lock is acquired, False otherwise
        - ltag: the lock tag
        - wtag: the write tag
        """
        if address not in self.locks:
            return False, -1, -1
        ret_val, ltag = self.locks[address].acquire_lock()

        # the lease seconds applies if the lock is acquired by a remote client
        # this client could potentially fail and keep the lock forever, thus
        # we release the lock after the lease_seconds
        if ret_val and lease_seconds is not None:
            # ensure thread safety
            def timer_callback(_ltag, _address):
                val, _ = self.locks[_address].release_lock(_ltag)
                if val:
                    print(f"[LOCK TIMER] lock released for address {_address}")
            th.Timer(lease_seconds, timer_callback, args=(ltag, address)).start()

        return ret_val, ltag, self.memory[address].wtag
    
    def release_lock(self, address: int, lease_ltag) -> tuple[bool, int, int]:
        """
        Return:
        - ret_val: True if the lock is released, False otherwise
        - ltag: the lock tag
        - wtag: the write tag
        """
        if address not in self.locks:
            return False, -1, -1
        wtag = self.memory[address].wtag
        ret_val, ltag = self.locks[address].release_lock(lease_ltag)
        return ret_val, ltag, wtag
    
    def set_status(self, address: int, status: str) -> bool:
        if address not in self.memory:
            return False
        self.memory[address].status = status
        return True

    def get_copy_holders(self, address: int) -> list[tuple[str, int]]:
        if address not in self.copy_holders:
            return []
        return self.copy_holders[address].copy()
    
    def add_copy_holder(self, address: int, holder: tuple[str, int]) -> bool:
        """
        Return:
        - True: if the holder is in the copy_holders list
        """
        if address not in self.copy_holders:
            return False
        if holder in self.copy_holders[address]:
            return True
        self.copy_holders[address].append(holder)
        self.memory[address].status = "S"
        return True
    
    def remove_copy_holder(self, address: int, holder: tuple[str, int]) -> bool:
        """
        Return:
        - True: if the holder is not in the copy_holders list
        """
        if address not in self.copy_holders:
            return False
        if holder not in self.copy_holders[address]:
            return True
        self.copy_holders[address].remove(holder)
        if len(self.copy_holders[address]) == 0:
            self.memory[address].status = "E"
        return True
    

File: java_code/edcs/src/main/edcs/project/MemoryManager.java
package main.edcs.project;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

public class MemoryManager {
    private Map<Integer, MemoryItem> memory;
    private Map<Integer, LockItem> locks;
    private Map<Integer, List<Tuple<String, Integer>>> copyHolders;
    private Tuple<Integer, Integer> memoryRange;
    private Timer lockTimer;

    public MemoryManager(Tuple<Integer, Integer> memoryRange) {
        this.memory = new ConcurrentHashMap<>();
        this.locks = new ConcurrentHashMap<>();
        this.copyHolders = new ConcurrentHashMap<>();
        this.memoryRange = memoryRange;
        // the lockTimer is basically a thread that is tasked with
        // running all the lock-lease timeout tasks that occur
        // when a remote process acquires a lock
        // This is an implementation difference compared to our python code
        this.lockTimer = new Timer();

        for(int i = this.memoryRange.getX(); i < this.memoryRange.getY(); i++) {
        	this.memory.put(i, new MemoryItem(null, "E"));
        	this.locks.put(i, new LockItem());
        	this.copyHolders.put(i, new ArrayList<>());
        }
    }

    public MemoryItem readMemory(int address) {
        return this.memory.getOrDefault(address, null);
    }

    public MemoryItem writeMemory(int address, Object data) {
        MemoryItem item = this.memory.get(address);
        
        if(item != null) {
            item.setData(data);
            item.setWtag(item.getWtag() + 1);
        }
        
        return item;
    }

    // return:
    // - retVal: true if the lock is acquired, false otherwise
    // - ltag: the lock tag
    // - wtag: the write tag
    public Tuple2<Boolean, Long, Long> acquireLock(int address, Long leaseSeconds) throws InterruptedException {
        LockItem lockItem = this.locks.get(address);
        
        if(lockItem == null) {
            return new Tuple2<>(false, (long) -1, (long) -1);
        }
        
        Tuple<Boolean, Long> lockResult = lockItem.acquireLock();
        boolean retVal = lockResult.getX();
        long ltag = lockResult.getY();

        // the lease seconds applies if the lock is acquired by a remote client
        // this client could potentially fail and keep the lock forever, thus
        // we release the lock after the lease_seconds
        if (retVal && leaseSeconds != null) {
            // class Timer is thread safe
            lockTimer.schedule(
                    new TimerTask() {
                        @Override
                        public void run() {
                            Tuple<Boolean, Long> resp = lockItem.releaseLock(ltag);
                            boolean success = resp.getX();
                            if (success) {
                                System.out.println("[LOCK TIMER] lock released for memory address: " + address);
                            }
                        }
                    }, leaseSeconds * 1000L
            );
        }

        return new Tuple2<>(retVal, ltag, this.memory.get(address).getWtag());
    }

    // return:
    // - retVal: true if the lock is released, false otherwise
    // - ltag: the lock tag
    // - wtag: the write tag
    public Tuple2<Boolean, Long, Long> releaseLock(int address, long leaseLtag) {
        LockItem lockItem = this.locks.get(address);
        
        if(lockItem == null) {
            return new Tuple2<>(false, (long) -1, (long) -1);
        }
        MemoryItem memoryItem = this.memory.get(address);
        long wtag = memoryItem.getWtag();
        Tuple<Boolean, Long> releaseResult = lockItem.releaseLock(leaseLtag);

        return new Tuple2<>(releaseResult.getX(), releaseResult.getY(), wtag);
    }

    public boolean setStatus(int address, String status) {
        MemoryItem memoryItem = this.memory.get(address);
        
        if(memoryItem == null) {
            return false;
        }
        
        memoryItem.setStatus(status);
        
        return true;
    }

    public List<Tuple<String, Integer>> getCopyHolders(int address) {
        return new ArrayList<>(this.copyHolders.getOrDefault(address, Collections.emptyList()));
    }

    // return:
    // - true: if the holder is in the copyHolders list
    public boolean addCopyHolder(int address, Tuple<String, Integer> holder) {
        List<Tuple<String, Integer>> holders = this.copyHolders.get(address);
        
        if(holders == null) {
            return false;
        }

        if (holders.contains(holder)) {
            return true;
        }
        
        holders.add(holder);
        this.memory.get(address).setStatus("S");
        
        return true;
    }

    // return:
    // - true: if the holder is not in the copyHolders list
    public boolean removeCopyHolder(int address, Tuple<String, Integer> holder) {
        List<Tuple<String, Integer>> holders = this.copyHolders.get(address);
        
        if(holders == null) {
            return false;
        }

        if (!holders.contains(holder)) {
            return true;
        }

        holders.remove(holder);
        
        if(holders.isEmpty()) {
        	this.memory.get(address).setStatus("E");
        }
        
        return true;
    }
}
